Add Forest dashboard

yilinxia opened the pull request:
add forest_dashboard including
- model selection
- Forest Plot
- Ridget Plot
- Add rope based model , data variable and coordinate variable

ndmlny-qs commented on arviz_dashboard/forest.py:
```diff
@@ -0,0 +1,217 @@
+                combined=True,
+                colors="cycle",
+            )
+            return forest_plt[0][0]
```
magic numbers again (`[0][0]`)

yilinxia replied:
this is an interesting question, so the `forest_plot` is a array as well as `forest_plot[0]`. Right now, I change it to `forest_plot.base[0]`

ndmlny-qs replied:
You are correct that it is a nested array. I think this is a feature from the matplotlib interface. You can leave a comment indicating what it is, and why you are indexing `[0][0]` to get the bokeh plot. I'm not sure `forest_plot.base[0]` is any clearer than `[0][0]`

ndmlny-qs commented on arviz_dashboard/forest.py:
```diff
@@ -0,0 +1,217 @@
+            op_slider,
+        ):
+            # calculate the ridgeplot_quantiles
+            temp_quant = list(ridge_quant)
+            quant_ls = temp_quant
+            quant_ls.sort()
```
why not `quantiles = sorted(ridge_quantiles)`

ndmlny-qs commented on arviz_dashboard/forest.py:
```diff
@@ -0,0 +1,217 @@
+
+            ridge_plt = az.plot_forest(
+                data,
+                model_names=multi_select,
+                kind="ridgeplot",
+                hdi_prob=thre_slider,
+                ridgeplot_truncate=truncate_checkbox,
+                ridgeplot_quantiles=quant_ls,
+                ridgeplot_overlap=op_slider,
+                backend="bokeh",
+                figsize=(9, 9),
+                show=False,
+                combined=True,
+                colors="white",
```
I'm a huge fan of ensuring the key/values exist in methods. Keep this pattern up everywhere.

ndmlny-qs commented on arviz_dashboard/forest.py:
```diff
@@ -0,0 +1,217 @@
+            get_ridge_plot,
+        )
+        # show up
+        display(
```
You are assuming this is run in a Juptyer environment. You should comment about that fact.

ahartikainen commented on arviz_dashboard/__init__.py:
```diff
@@ -3,5 +3,6 @@
 
 from .elpd import dashboard_elpd
 from .ppc import dashboard_ppc
+from .forest import *
```
I think in `__init__`, the use of `from .xxx import *` is normal, if `__all__` is defined in xxx

ndmlny-qs replied:
True. I think a discussion around a style guide for the dashboard repo is in order, but for now @yilinxia you can resolve this conversation as done after you read it without any changes, as a change is out of scope for this PR.

I personally find the import style from the Google Python style guide useful as I've encountered some confusing bugs when `from x import *` is used. The style guide can be found here: https://google.github.io/styleguide/pyguide.html#22-imports. Also `xarray` mentions not using wildcard imports here: (https://github.com/pydata/xarray/issues/1080#issuecomment-259270684) for readability purposes, which I agree with.

Counter arguments to why we should not care are welcome. Perhaps in a new discussion about style guides.

ahartikainen replied:
Are we now talking about

    from x import *

or

    from .x import *

yilinxia commented on arviz_dashboard/forest.py:
```diff
@@ -0,0 +1,217 @@
+        else:
+            coor_variables = [""]
+        self.param["coor_variable"].objects = coor_variables
+        if self.coor_variable not in coor_variables:
+            self.coor_variable = coor_variables[0]
```
these two lines can be removed

ndmlny-qs commented:
We need to update the notebook to save the log likelihood in the inference data object. We should also pin panel in the requirements file to be less than 1 so we get the right version of bokeh installed.

@yilinxia I think we can merge after those changes.

yilinxia commented:
@ndmlny-qs Hi Andy, I have addressed the issues you pointed out. Meanwhile, I am also thinking about maybe we should use the same standard to “beautify” the code I wrote before (in another branch?)

ndmlny-qs commented:
I agree, we need to have steps that incorporate `pre-commit` for linting, typing and formatting; and ensure those items are done in the GitHub actions. I would say (for now) that we can squash and merge this, and work on a dev protocol in the readme. We can use some of what is talked about in ArviZ's contributing page as a guide

ndmlny-qs reviewed (approved):
lgtm

yilinxia commented:
@aloctavodia Hi Osvaldo, do you any other comments? otherwise, I think we can merge this branch
